### DISCUSSION TITLE
GPT NeoX/StabilityAI StableLM support

### DISCUSSION THREAD

**Discussion by rabidcopy** [LostRuins/koboldcpp#93] [Ideas]
A bit of a cross post from llama.cpp but feels appropriate here as this repo targets a number of ggml formats instead of just LLaMa. https://github.com/ggerganov/llama.cpp/issues/1063 It might not be too out of the question as someone has already done work on converting/quantizing GPT NeoX models in ggml. Strangely none of this was PR'd to ggml and has just been out there.
https://github.com/NolanoOrg/cformers 
https://github.com/NolanoOrg/cformers/blob/master/cformers/cpp/converters/convert_gptneox_to_ggml.py 
https://github.com/NolanoOrg/cformers/blob/master/cformers/cpp/quantize_gptneox.cpp
Edit: https://github.com/ggerganov/ggml/pull/96

**Comment by LostRuins** [comment]
Hmm I'd have to take a look - the version of ggml they are using is **ancient** and will not be directly compatible with the existing library. Also on the surface it looks like NeoX is very similar to GPT-J just with a different RoPE function, but i'm not too familiar with the specifics yet. 

Ideally ggerganov does a sample implementation compatible with the latest GGML and I can just use theirs. We'll see how that develops before I try to roll my own version.

**Comment by rabidcopy** [comment]
@LostRuins ggml implementation was merged https://github.com/ggerganov/ggml/pull/96 so it might be in a more workable state to implement here.

**Reply by LostRuins** [reply]
Yep that's the plan, after I resolve the cublas stuff either by merging or dropping it.
